frameworks/keyed/million/src/main.js

```diff
@@ -0,0 +1,336 @@
+let main;
+
+const clear = () => {
+  list = [];
+  main.remove();
+};
+
+const buildData = (count) => {
+  const data = new Array(count);
+  for (let i = 0; i < count; ++i) {
+    data[i] = {
+      id: nextId++,
+      label: `${adjectives[random(adjectives.length)]} ${
+        colors[random(colors.length)]
+      } ${nouns[random(nouns.length)]}`,
+    };
+  }
+  return data;
+};
+
+const create1k = () => {
+  if (list.length) clear();
+  list = buildData(1000);
+  update();
+  return false;
+};
+
+const create10k = () => {
+  if (list.length) clear();
+  list = buildData(10000);
+  update();
+  return false;
+};
+
+const append1k = () => {
+  list = list.concat(buildData(1000));
+  update();
+  return false;
+};
+
+const updateEvery10 = () => {
+  let i = 0;
+  while (i < list.length) {
+    list[i].label = `${list[i].label} !!!`;
+    i += 10;
+  }
+  update();
+  return false;
+};
+
+const swapRows = () => {
+  if (list.length > 998) {
+    const item = list[1];
+    list[1] = list[998];
+    list[998] = item;
+  }
+  update();
+  return false;
+};
+
+let prevBlock;
+
+const select = (id) => {
+  if (prevBlock) {
+    if (prevBlock.props.id === id) return;
+    const { id: prevId, label } = prevBlock.props;
+    prevBlock.patch(
+      Row(
+        {
+          id: prevId,
+          label,
+          className: '',
+        },
+        String(id)
+      )
+    );
+  }
+  const block = main.children.find((block) => block.props.id === id);
+  const { label } = block.props;
+  block.patch(
+    Row(
+      {
+        id,
+        label,
+        className: 'danger',
+      },
+      String(id)
+    )
+  );
+
+  prevBlock = block;
+};
+
+const remove = (id) => {
+  const index = list.findIndex((item) => item.id === id);
+  list.splice(index, 1);
+  main.children[index].remove();
+  main.children.splice(index, 1);
+};
+
+const Row = (() => {
+  const root = stringToDOM(
+    '<tr><td class="col-md-1"></td><td class="col-md-4"><a></a></td><td class="col-md-1"><a><span class="glyphicon glyphicon-remove" aria-hidden="true"></span></a></td><td class="col-md-6"></td></tr>'
+  );
+  const edits = [
+    {
+      path: [0],
+      edits: [
+        {
+          type: 'child',
+          hole: 'id',
+          index: 0,
+        },
+      ],
+      inits: [],
+    },
+    {
+      path: [1, 0],
+      edits: [
+        {
+          type: 'child',
+          hole: 'label',
+          index: 0,
+        },
+      ],
+      inits: [],
+    },
+    {
+      path: [],
+      edits: [
+        {
+          type: 'attribute',
+          hole: 'className',
+          name: 'class',
+        },
+      ],
+      inits: [],
+    },
+  ];
+  const shouldUpdate = (oldProps, newProps) => {
+    return (
+      oldProps.label !== newProps.label ||
+      oldProps.className !== newProps.className
+    );
+  };
+  return (props, key) => {
+    return new Block(root, edits, props, key, shouldUpdate);
+  };
+})();
+
+(() => {
+  new Block(
+    stringToDOM(
```

Comment:
Does this count as too much optimization here?

Reply:
Is this code generated by a compiler? (Asking since it's under /src)

Reply:
No, it's not. I can rewrite it such that it uses the canonical API instead, do you think that's more aligned with this bench?

Reply:
Yeah that would probably be better, I can't imagine people in practice will write code like that.

Reply:
Got it, thanks!